allennlp/models/constituency_parser.py

```diff
@@ -285,169 +419,36 @@ def construct_trees(self,
             in the batch.
+        num_trees: ``int``, required.
+            The number of trees to be returned for each sentence.
 
         Returns
         -------
         A ``List[Tree]`` containing the decoded trees for each element in the batch.
+        :param sentences:
+        :param predictions:
+        :param all_spans:
+        :param num_trees:
         """
         # Switch to using exclusive end spans.
-        exclusive_end_spans = all_spans.clone()
+        exclusive_end_spans = all_spans.clone().cpu().numpy()
         exclusive_end_spans[:, :, -1] += 1
-        no_label_id = self.vocab.get_token_index("NO-LABEL", "labels")
-
-        trees: List[Tree] = []
-        for batch_index, (scored_spans, spans, sentence) in enumerate(zip(predictions,
-                                                                          exclusive_end_spans,
-                                                                          sentences)):
-            selected_spans = []
-            for prediction, span in zip(scored_spans[:num_spans[batch_index]],
-                                        spans[:num_spans[batch_index]]):
-                start, end = span
-                no_label_prob = prediction[no_label_id]
-                label_prob, label_index = torch.max(prediction, -1)
-
-                # Does the span have a label != NO-LABEL or is it the root node?
-                # If so, include it in the spans that we consider.
-                if int(label_index) != no_label_id or (start == 0 and end == len(sentence)):
-                    # TODO(Mark): Remove this once pylint sorts out named tuples.
-                    # https://github.com/PyCQA/pylint/issues/1418
-                    selected_spans.append(SpanInformation(start=int(start), # pylint: disable=no-value-for-parameter
-                                                          end=int(end),
-                                                          label_prob=float(label_prob),
-                                                          no_label_prob=float(no_label_prob),
-                                                          label_index=int(label_index)))
-
-            # The spans we've selected might overlap, which causes problems when we try
-            # to construct the tree as they won't nest properly.
-            consistent_spans = self.resolve_overlap_conflicts_greedily(selected_spans)
-
-            spans_to_labels = {(span.start, span.end):
-                               self.vocab.get_token_from_index(span.label_index, "labels")
-                               for span in consistent_spans}
-            sentence_pos = pos_tags[batch_index] if pos_tags is not None else None
-            trees.append(self.construct_tree_from_spans(spans_to_labels, sentence, sentence_pos))
 
+        trees: List[List[Tree]] = []
+        for batch_index in range(len(sentences)):
+            span_to_index = {}
+            for span_index in range(num_spans[batch_index]):
```

Comment:
Can you change this iteration over the batch and spans back to the previous method using `enumerate`? This way, a reader has to know what each of the dimensions of the various tensors represent, whereas previously they had names.